src/planning/src/planning/path_smoothing.cpp

```diff
@@ -0,0 +1,220 @@
+  X = gsl_matrix_alloc(n, ncoeffs);
+  Y = gsl_matrix_alloc(n, ncoeffs);
+  c = gsl_vector_alloc(ncoeffs);
+  c2 = gsl_vector_alloc(ncoeffs);
+  w = gsl_vector_alloc(n);
+  cov = gsl_matrix_alloc(ncoeffs, ncoeffs);
+  cov2 = gsl_matrix_alloc(ncoeffs, ncoeffs);
+  mw = gsl_multifit_linear_alloc(n, ncoeffs);
+  mw2 = gsl_multifit_linear_alloc(n, ncoeffs);
+
+
+  // Set spline data
+  for (int i = 0; i < n; i++) {
+    gsl_vector_set(i_values, i, i);
+    gsl_vector_set(x_values, i, path[i]->getX());
+    gsl_vector_set(y_values, i, path[i]->getY());
+    // closer cones more important(better stability)
+    gsl_vector_set(w, i, 1.0 / pow(path[i]->getDistanceTo(path[(i + 1) % n]), 2));
+  }
+
+  // Set i range within cone set length
+  gsl_bspline_knots_uniform(0, n, bw);
+  gsl_bspline_knots_uniform(0, n, cw);
+
+  /* construct the fit matrix X */
+  for (int i = 0; i < n; i++) {
+    /* compute B_j(xi) for all j */
+    gsl_bspline_eval(i, B, bw);
+    gsl_bspline_eval(i, C, cw);
+
+    /* fill in row i of X */
+    for (int j = 0; j < ncoeffs; j++) {
+      double Bj = gsl_vector_get(B, j);
+      gsl_matrix_set(X, i, j, Bj);
+      double Cj = gsl_vector_get(C, j);
+      gsl_matrix_set(Y, i, j, Cj);
+    }
+  }
+
+  // Fit spline
+  gsl_multifit_wlinear(X, w, x_values, c, cov, &chisq, mw);
+  gsl_multifit_wlinear(Y, w, y_values, c2, cov2, &chisq2, mw2);
+
+  //Initialize variables for subsequent spline evaluation
+  double xi, yi, yerr, yerr2;
+  std::vector<double> i_eval, x_eval, y_eval;
+  std::vector<PathPoint *> path_eval;
+
+  // Calculate the desired amount of spline points and add them to "path_eval"
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < precision; j += 1) {  // iterate over decimal numbers
+      gsl_bspline_eval(i + static_cast<float>(j) / precision, B, bw);
+      gsl_bspline_eval(i + static_cast<float>(j) / precision, C, cw);
+      gsl_multifit_linear_est(B, c, cov, &xi, &yerr);
+      gsl_multifit_linear_est(C, c2, cov2, &yi, &yerr2);
+      i_eval.push_back(i);
+      x_eval.push_back(xi);
+      y_eval.push_back(yi);
+      PathPoint *path_point = new PathPoint(xi, yi);
+      path_eval.push_back(path_point);
+      if (j == 0 && i == n - 1) {
+        break; // Decimals can't go over last int
+      }
+    }
+  }
+
+  // Free Memory
+  gsl_bspline_free(bw);
+  gsl_bspline_free(cw);
+  gsl_vector_free(B);
+  gsl_vector_free(C);
+  gsl_vector_free(i_values);
+  gsl_vector_free(x_values);
+  gsl_vector_free(y_values);
+  gsl_matrix_free(X);
+  gsl_matrix_free(Y);
+  gsl_vector_free(c);
+  gsl_vector_free(w);
+  gsl_matrix_free(cov);
+  gsl_multifit_linear_free(mw);
+  gsl_multifit_linear_free(mw2);
+
+  RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"),
+  "END fitSpline with %i points", static_cast<int>(path_eval.size()));
+
+  return path_eval;
+}
+
+std::vector<PathPoint *> PathSmoothing::pathSmoother(int precision,
+    int order, float coeffs_ratio, std::vector<PathPoint *> unord_path) {
```

Comment:
Any reason for this method to exist? It seems a bit of excessive delegation to have validate call this that calls spline smoother don't you think?

Reply:
The reason behind having "validate" as an independent method is to call this function with predefined values for the first parameters without actually defining a default value for them since there's nothing special about any value, while also leaving the possibility to call "pathSmoother" with other inputs.

The reason behind having "pathSmoother" as an independent method is to maintain "orderPath" and "splineSmoother" as separate functions (increasing modularity and readability, but also leaving a chance to change only one of them) but to have a way to call them together since that's the only way in which they should be used.